### PR TITLE
Docker: Component based Multi stage docker

### PR DESCRIPTION
**Pull request by heliocastro**
See individual commits for details.

### CODE REVIEW

**Comment by tsteenbe** [comment]
@heliocastro  As discussed on ORT Slack made some commit message improvement such as spelling, grammar, etc. Believe we are now good to go.

**Comment by sschuberth** [comment]
> Believe we are now good to go.

Sorry, still "no" from my side. Many of @mnonnenmacher's [very good points](https://github.com/oss-review-toolkit/ort/pull/4746#issuecomment-1241122895) still hold with this PR. Even with the reduced number of commits, unrelated stuff if done in  single commit, formatting does not adhere to our standards, etc.

I believe the way forward would be as @mnonnenmacher described:

> * One commit to copy the current Dockerfile to the legacy directory.
> 
> * A few commits that apply the relevant changes on top of the current Dockerfile, but without all the version updates that just repeat version updates which already happened in our current Dockerfile. Or even squashing all changes into a single commit would be better than the current state of the PR.
> 
> * One commit for each version update that changes a version compared to our current Dockerfile.

File: Dockerfile [hunk 1]
```diff
@@ -19,41 +20,112 @@
 # SPDX-License-Identifier: Apache-2.0
 # License-Filename: LICENSE
 
-# Set this to the version ORT should report.
-ARG ORT_VERSION="DOCKER-SNAPSHOT"
+#------------------------------------------------------------------------
+# Use OpenJDK Eclipe Temurin Ubuntu LTS
+FROM eclipse-temurin:11-jdk AS build
```
Comment (sschuberth) [inline Dockerfile]: We've already used `eclipse-temurin:11-jdk-jammy` before, so "Base main Docker on OpenJDK 11 Eclipse Temurin" in the commit message is misleading.

File: Dockerfile [hunk 2]
```diff
@@ -153,6 +153,23 @@ RUN curl -ksSL https://github.com/CocoaPods/CocoaPods/archive/9461b346aeb8cba6df
 
 COPY docker/ruby.sh /etc/ort/bash_modules
 
+#------------------------------------------------------------------------
+# SCANCODE - Build scancode as a separate component
+FROM pythonbuild AS scancodebuild
+
+ARG SCANCODE_VERSION="3.2.1rc2"
```
Comment (sschuberth) [inline Dockerfile]: This is a step backwards. We use version 30.1.0 already in the current `Dockerfile`.

File: Dockerfile [hunk 3]
```diff
@@ -257,15 +279,7 @@ RUN /opt/ort/bin/import_proxy_certs.sh && \
...
-        # Trigger ScanCode configuration for Python 3 and reindex licenses initially.
-        cd /usr/local/scancode-toolkit-$SCANCODE_VERSION && \
-        PYTHON_EXE=/usr/bin/python3 /usr/local/scancode-toolkit-$SCANCODE_VERSION/scancode --reindex-licenses && \
-        chmod -R o=u /usr/local/scancode-toolkit-$SCANCODE_VERSION && \
-        ln -s /usr/local/scancode-toolkit-$SCANCODE_VERSION/scancode /usr/local/bin/scancode
-
+    yes | $ANDROID_HOME/cmdline-tools/bin/sdkmanager $SDK_MANAGER_PROXY_OPTIONS --sdk_root=$ANDROID_HOME "platform-tools"
 
 ENTRYPOINT ["/usr/bin/ort"]
+
```
Comment (sschuberth) [inline Dockerfile]: Superfluous change.

File: Dockerfile [hunk 4]
```diff
@@ -67,8 +139,8 @@ ENV \
     # SDK versions.
     ANDROID_SDK_VERSION=6858069 \
     # Installation directories.
-    ANDROID_HOME=/opt/android-sdk \
-    GOPATH=/tmp/go
+    ANDROID_HOME=/opt/android-sDockerfiledk \
```
Comment (sschuberth) [inline Dockerfile]: Typo in the path. This could have never worked.

File: .dockerignore [hunk 5]
```diff
@@ -1,6 +1,5 @@
 # Ignore the Docker files themselves to avoid that all layers are invalidated if
 # the Docker files are changed during development.
-.dockerignore
```
Comment (sschuberth) [inline .dockerignore:3]: This should not be removed. See the comment above.

File: Dockerfile [hunk 6]
```diff
@@ -34,67 +34,77 @@ ARG CRT_FILES=""
 COPY "$CRT_FILES" /tmp/certificates/
 RUN /etc/scripts/import_proxy_certs.sh \
     && if [ -n "$CRT_FILES" ]; then \
-        /etc/scripts/import_certificates.sh /tmp/certificates/; \
-       fi
+    /etc/scripts/import_certificates.sh /tmp/certificates/; \
```
Comment (sschuberth) [inline Dockerfile:37]: Indentation is still broken after this change.

File: Dockerfile [hunk 7]
```diff
@@ -34,67 +34,77 @@ ARG CRT_FILES=""
...
-        libssl-dev \
-        make \
-        netbase \
-        tk-dev \
-        tzdata \
-        uuid-dev \
-        unzip \
-        xz-utils \
-        zlib1g-dev \
+    build-essential \
```
Comment (sschuberth) [inline Dockerfile:46]: Broken indentation.

File: .gitignore [hunk 8]
```diff
@@ -12,3 +12,7 @@ out/
 
 # Mac
 .DS_Store
+
+# Ignore node_modules
```
Comment (sschuberth) [inline .gitignore:16]: These changes seem unrelated to anything mentioned in the commit message.

File: scripts/docker_build.sh [hunk 9]
```diff
@@ -23,10 +23,12 @@
 GIT_ROOT=$(git rev-parse --show-toplevel)
 GIT_VERSION=$(git describe --abbrev=10 --always --tags --dirty)
 ORT_DOCKER=${ORT_DOCKER:-ort}
+ARCH_TRANSLATION=""
 
 echo "Setting ORT_VERSION to $GIT_VERSION."
-DOCKER_BUILDKIT=1 docker build -f "$GIT_ROOT"/Dockerfile \
+docker buildx build -f "$GIT_ROOT"/Dockerfile \
```
Comment (sschuberth) [inline scripts/docker_build.sh:29]: Switching to `buildx` is a rather big under-the-hood change that deserves its own commit with a thorough explanation.

**Review by sschuberth — CHANGES_REQUESTED** [review]
Sorry, still way too messy.

File: Dockerfile (hunk 1 above)
Comment (tsteenbe) [reply Dockerfile]: No longer applicable as commit was squashed.

File: docker/legacy/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,177 @@
+# syntax=docker/dockerfile:1.3
```
Comment (sschuberth) [inline docker/legacy/Dockerfile:1]: Commit message nits:

- "to new multi stage Dockerfile" lacks a "to *the* new" or "to *a* new".
- please don't emphasize "multi stage"; that's confusing because the legacy `Dockerfile` also is multi-stage already.

I propose: "This is a preparation for allowing ORT users to stick to the current Dockerfile instead of having to directly upgrade to the upcoming rewritten Dockerfile."

File: scripts/docker_build.sh [hunk 11]
```diff
@@ -20,5 +21,10 @@
 GIT_ROOT=$(git rev-parse --show-toplevel)
 GIT_REVISION=$(git describe --abbrev=10 --always --tags --dirty)
 
-echo "Setting ORT_VERSION to $GIT_REVISION."
-DOCKER_BUILDKIT=1 docker build -f $GIT_ROOT/Dockerfile -t ort --build-arg ORT_VERSION=$GIT_REVISION $GIT_ROOT
+echo "Setting ORT_VERSION to $GIT_VERSION."
```
Comment (sschuberth) [inline scripts/docker_build.sh:28]: This should still say `GIT_REVISION`.

File: scripts/docker_build.sh [hunk 12]
```diff
@@ -20,5 +21,10 @@
...
+docker buildx build -f "$GIT_ROOT"/Dockerfile \
```
Comment (sschuberth) [inline scripts/docker_build.sh:29]: Nit: I'd just quote all of the path for clarity.

File: scripts/docker_build.sh [hunk 13]
```diff
@@ -20,5 +21,10 @@
...
+    -t "$ORT_DOCKER" \
```
Comment (sschuberth) [inline scripts/docker_build.sh:30]: There is no `ORT_DOCKER` variable; so this should probably stay at "ort".

File: scripts/docker_build.sh [hunk 14]
```diff
@@ -20,5 +21,10 @@
...
+    --build-arg ORT_VERSION="$GIT_VERSION" \
```
Comment (sschuberth) [inline scripts/docker_build.sh:31]: Again, this should still say GIT_REVISION.

File: scripts/docker_build.sh [hunk 15]
```diff
@@ -20,5 +21,10 @@
...
+    --platform linux/amd64 \
+    "$@" \
+    "$GIT_ROOT"
```
Comment (sschuberth) [inline scripts/docker_build.sh:34]: Ideally, all the quoting an wrapping that you do here should go to separate commits, so it becomes apparent that basically the only thing that this commit does is to add "buildx" and "--platform linux/amd64".

File: scripts/docker_build.sh [hunk 16]
```diff
@@ -20,5 +21,10 @@
...
 GIT_REVISION=$(git describe --abbrev=10 --always --tags --dirty)
 
-echo "Setting ORT_VERSION to $GIT_REVISION."
-DOCKER_BUILDKIT=1 docker build -f $GIT_ROOT/Dockerfile -t ort --build-arg ORT_VERSION=$GIT_REVISION $GIT_ROOT
+echo "Setting ORT_VERSION to $GIT_VERSION."
+docker buildx build -f "$GIT_ROOT"/Dockerfile \
+    -t "$ORT_DOCKER" \
+    --build-arg ORT_VERSION="$GIT_VERSION" \
+    --platform linux/amd64 \
+    "$@" \
```
Comment (sschuberth) [inline scripts/docker_build.sh:33]: This is new. Why is it needed?

File: scripts/docker_build.sh [hunk 17]
```diff
@@ -1,6 +1,7 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (sschuberth) [inline scripts/docker_build.sh:1]: What Bash feature is being used now?

File: docker/00-add_local_path.sh [hunk 18]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (sschuberth) [inline docker/00-add_local_path.sh:1]: Commit message: Again, please don't emphasize "multi stage". It has been multi-stage before.

File: scripts/docker_build.sh [hunk 19]
```diff
@@ -18,8 +18,12 @@
 # SPDX-License-Identifier: Apache-2.0
 # License-Filename: LICENSE
 
+# Docker from ORT uses BuildKit feature
+
 GIT_ROOT=$(git rev-parse --show-toplevel)
-GIT_REVISION=$(git describe --abbrev=10 --always --tags --dirty)
+GIT_VERSION=$(git describe --abbrev=10 --always --tags --dirty)
+ORT_DOCKER=${ORT_DOCKER:-ort}
+ARCH_TRANSLATION=""
```
Comment (sschuberth) [inline scripts/docker_build.sh:26]: These changes (renaming the `GIT_REVISION` variable, making the Docker tag to use configurable, etc.) are unrelated to the topic described in the commit message.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scripts/docker_build.sh (hunk 13 above)
Comment (heliocastro) [reply scripts/docker_build.sh:30]: If someone pass ORT_DOCKER=ort_my_custom the container will be created  as ort_my_custom, otherwise will be used as ort

File: scripts/docker_build.sh (hunk 16 above)
Comment (heliocastro) [reply scripts/docker_build.sh:33]: Build on Arm
Ask @tsteenbe

File: scripts/docker_build.sh (hunk 13 above)
Comment (heliocastro) [reply scripts/docker_build.sh:29]: So let's say you want a custom image and want specia version of python and want to keep the pristine docker image ?
`PYTHON_VERSION=3.9.9 ORT_DOCKER=ort_python399 ./scripts/docker_build.sh`

File: scripts/docker_build.sh (hunk 16 above)
Comment (tsteenbe) [reply scripts/docker_build.sh:32]: The `--platform linux/amd64`  forces Docker to use linux/amd64 platform by default on Apple Silicon.

File: batect.yml [hunk 20]
```diff
@@ -94,4 +94,5 @@ tasks:
       - buildFunTest
     run:
       container: run
-      command: ./gradlew funTest
+      command: "'./gradlew funTest jacocoFunTestReport -p analyzer'"
```
Comment (sschuberth) [inline batect.yml:97]: Can you please explain what this does / why it is necessary? The commit message unfortunately does not give me any clue. In particular:

- Why are JaCoCo reports being created now?
- Why is running the tests now limited to functional tests in the analyzer?

Comment (tsteenbe) [reply batect.yml]: @heliocastro removed `jacocoFunTestReport -p analyzer` but now there is below error.

@sschuberth  Any clue what is causing this?

```

org.ossreviewtoolkit.model.utils.PostgresFileArchiverStorageFunTest > Before Spec Error FAILED
    java.lang.IllegalStateException: Could not find a valid Docker environment. Please see logs and check configuration
        at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$getFirstValidStrategy$4(DockerClientProviderStrategy.java:156)
        at java.base/java.util.Optional.orElseThrow(Optional.java:408)
        at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:148)
        at org.testcontainers.DockerClientFactory.getOrInitializeStrategy(DockerClientFactory.java:146)
        at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:188)
        at org.testcontainers.DockerClientFactory$1.getDockerClient(DockerClientFactory.java:101)
        at com.github.dockerjava.api.DockerClientDelegate.authConfig(DockerClientDelegate.java:107)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:316)
        at com.opentable.db.postgres.embedded.EmbeddedPostgres.<init>(EmbeddedPostgres.java:99)
        at com.opentable.db.postgres.embedded.EmbeddedPostgres$Builder.start(EmbeddedPostgres.java:353)
        at org.ossreviewtoolkit.utils.test.PostgresListener.beforeSpec(PostgresListener.kt:42)
        at io.kotest.engine.spec.SpecExtensions.beforeSpec-gIAlu-s(SpecExtensions.kt:49)
        at io.kotest.engine
```

**Comment by fviernau** [comment]
I've reviewed this again (not everything in most possible detail), haven't found any blockers.
I've compared the tool versions and the upgrades now  do happen in separate commits, e.g. the large commit should keep the versions identical. The names of some variables change, like e.g. `ANDROID_SDK_VERSION` -> `ANDROID_CMD_VERSION`, but in this case I believe this isn't an issue at all.

File: batect.yml (hunk 20 above)
Comment (sschuberth) [reply batect.yml]: I don't have a concrete idea, but it looks like some of the changes broke running Docker-in-Docker, as our `PostgresFileArchiverStorageFunTest` relies on `org.testcontainers`.

**Comment by sschuberth** [comment]
> I've reviewed this again (not everything in most possible detail), haven't found any blockers.

To me the last commit is still problematic as it touches the `batect.yml` file, which is exactly the bit I'd expect *not* to get touched in order to prove that the new `Dockerfile` does not break any test run via Batect WRT the current `Dockerfile`. Also, to me the commit message is not clear enough about what the change is actually needed for.

**Review by tsteenbe — APPROVED** [review]

File: scripts/docker_build.sh (hunk 17 above)
Comment (sschuberth) [reply scripts/docker_build.sh]: Why was this resolved while being unanswered?

File: scripts/docker_build.sh (hunk 11 above)
Comment (sschuberth) [reply scripts/docker_build.sh]: Still broken.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scripts/docker_build.sh (hunk 17 above)
Comment (heliocastro) [reply scripts/docker_build.sh]: Is already fixed.

**Comment by tsteenbe** [comment]
@sschuberth "Why was this resolved while being unanswered?" The current version of docker_build.sh is using  "#!/bin/sh" and not "#!/bin/bash" - you looking at outdated version for your code review I guess.

**Comment by heliocastro** [comment]
> > I've reviewed this again (not everything in most possible detail), haven't found any blockers.
> 
> To me the last commit is still problematic as it touches the `batect.yml` file, which is exactly the bit I'd expect _not_ to get touched in order to prove that the new `Dockerfile` does not break any test run via Batect WRT the current `Dockerfile`. Also, to me the commit message is not clear enough about what the change is actually needed for.

Container has several tool version managers that install base is not on standard PATH.
To made this available, was necessary add bash profiles to made visible to Ort. Profiles are only parsed if you run in login environment, and container by default not do that.

For Ort binary we use a wrapper where it calls bash -l so pushing in login environment

In batect, the case is that we calling gradle, not ort wrapper, meaning not bash  login shell, though failing to access tooling.
As documentation of batect https://batect.dev/docs/reference/config/containers/#entrypoint, this is correct way to execute a command from entry point shell.

As the jacoco move to container call, before was working due be executed as root user, and ignoring the fact that the second call passed was running as github actions user.
With the new container running things as a regular user **ort**, the process ends up as a ghaction user, which is different from **ort** user trying to write in a non permissive space

So moving the jacoco to inside container execution and not outside produce the correct results, as entire build and test are placed in same environment, same user and non privileged user.

File: scripts/docker_build.sh (hunk 17 above)
Comment (sschuberth) [reply scripts/docker_build.sh]: If so, then not in the correct commit. See how the current https://github.com/oss-review-toolkit/ort/pull/5760/commits/02787e58c054c34fe1ca07f2017131e0c2347212 still changes this line.

Comment (sschuberth) [reply scripts/docker_build.sh]: Yeah, you're changing it back only in https://github.com/oss-review-toolkit/ort/pull/5760/commits/07c68b664a7d2b9b1d7c367e4812feadafe8d677. What a mess...

Comment (tsteenbe) [reply scripts/docker_build.sh]: I have fixed it now the "docker_build: Enable builds on Apple Silicon" no longer changes line 1.

File: Dockerfile [hunk 21]
```diff
@@ -19,159 +20,335 @@
...
+COPY --from=rubybuild /etc/profile.d/ruby.sh /etc/profile.d/
+RUN chmod o+rwx ${RBENV_ROOT}
+
+# NodeJS
+ARG NODEJS_VERSION=16.17.0
+ARG NVM_DIR=/opt/nvm
+ENV NODE_PATH $NVM_DIR/v$NODEJS_VERSION/lib/node_modules
+ENV PATH $NVM_DIR/versions/node/v$NODEJS_VERSION/bin:$PATH
+COPY --chown=$USERNAME:$USERNAME --from=nodebuild ${NVM_DIR} ${NVM_DIR}
+RUN chmod o+rwx ${RBENV_ROOT}
```
Comment (nnobelis) [inline Dockerfile:298]: This should be `NVM_DIR`.

**Review by nnobelis — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 21 above)
Comment (heliocastro) [reply Dockerfile]: Fixed

**Comment by sschuberth** [comment]
Superseded by https://github.com/oss-review-toolkit/ort/pull/5830.
